### ISSUE TITLE
Should not allow views of NA-masked arrays which change the array size (migrated from Trac #2029)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#3576]
Original ticket http://projects.scipy.org/numpy/ticket/2029
Reported 2012-01-25 by atmention:mwiebe, assigned to unknown.

Chuck posted an example:

```
Looks like views of masked arrays have other problems:

In [13]: a = ones(3, int16, maskna=1)

In [14]: a.view(int8)
Out[14]: array([1, 0, 1, NA, 1, NA], dtype=int8)
```

This looks like a serious bug to me, to avoid memory corruption issues it should raise an exception.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2012-02-23

I think this changes the spirit of NumPy 1.X.    I don't know what the "memory corruption" issues are exactly you are referring to, but if someone knows what they are doing they can change the way memory is viewed and certainly alter it.   

To me it feels inconsistent to change this for masked arrays.   But, I'm not really sure I understand what is happening right now when you say maskna=1.

-Travis

**Comment by thouis** [comment]
Comment in Trac by atmention:mwiebe, 2012-02-23

Here's an example:

```
a = np.arange(3, maskna=True)
# This array has two things now, 3 int32s, and 3 corresponding mask elements
b = a.view(np.int16)
# This array has 6 int16s, but is supposed to share the 3 corresponding mask elements
# Right now it operates as if there are 6 elements, allowing reading/writing
# of invalid memory.
```

**Comment by thouis** [comment]
Comment in Trac by trac user bryan, 2012-02-27

Travis, are you on board with the proposed changes? There is a PR to to implement this strictly, but I plan to loosen it up some to allow changes in dtype when the dtypes have the same size.

**Comment by thouis** [comment]
Comment in Trac by atmention:teoliphant, 2012-02-27

It looks like anything that would change the number of mask locations should be disallowed.   But, it should just check for the size of the resulting mask but otherwise allow dtype modifications. 

-Travis

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2012-03-03

Fix pushed in 09f01d8.
